### ISSUE TITLE
format: 'json:results.${process.pid}.json' not unique enough

### ISSUE THREAD

**Issue by wswebcreation** [protractor-cucumber-framework/protractor-cucumber-framework#99]
I'm currently trying to find a new way to use the JSON formatter with CucumberJS 3 and thought I would use the provided `format: 'json:results.json'` and the PID that is provided by `protractor-cucumber-framework` if multiple feature-files are executed [see here](https://github.com/protractor-cucumber-framework/protractor-cucumber-framework#formatters-when-tests-are-sharded-or-with-multi-capabilities).

During some tests I saw that I didn't got the correct result in the generated JSON. I created some tests and put the result in the table below.

This table uses or the `capability` or the `multiCapabilities` from Protractor. There are 3 situations where the PID in the filename is not unique enough which results in clobbering each other and only showing the last run feature in the saved JSON file. This even gets worse when you have for example 10 feature-files and 2 instances to be used to run on. You will only see 2 report files of the last 2 run features because there are only 2 instances with a unique PID.

I hope this is clear enough. I'm currently also thinking about a solution for a PR, but if you have an idea it would also be nice.

| capability | multiCapabilities | shardTestFiles | maxInstances | features | json files | PID in filename | features in json | Remarks                                          |
|------------|-------------------|----------------|--------------|----------|------------|-----------------|------------------|--------------------------------------------------|
| true       |                   | false          |              | 1        | 1          | false           | 1                | success                                          |
| true       |                   | true           |              | 1        | 1          | true            | 1                | success                                          |
| true       |                   | false          |              | > 1      | 1          | false           | 2                | success                                          |
| **true**       |                   | **true**           | **1**            | **> 1**      | **1**          | **true**            | **1**              | **X , it will override the previous feature result** |
| true       |                   | true           | 2            | > 1      | 2          | true            | 1                | success                                          |
|            | true              | false          |              | 1        | 1          | true            | 1                | success                                          |
|            | true              | true           |              | 1        | 1          | true            | 1                | success                                          |
|            | true              | false          |              | > 1      | 1          | true            | 2                | success                                          |
|            | **true**              | **true**           |              | **> 1**      | **1**          | **true**            | **1**                | **X , it will override the previous feature result** |
|            | **true**              | **true**           | **1**            | **> 1**      | **1**          | **true**            | **2**                | **X , it will override the previous feature result** |
|            | true              | true           | 2            | > 1      | 2          | true            | 1                | success                                          |

**Comment by darrinholst** [comment]
I'm having trouble wrapping my head around why the process id wouldn't work. That would imply that cucumber itself is overwriting it. I updated my example app for your first failure case...sharded, multi feature, max instance 1. I get an output file for each feature file since protractor shards by file. What are you expecting the output to be in this case? 

https://github.com/dholst/protractor-cucumber-framework-example/blob/master/protractor.conf.js

**Comment by wswebcreation** [comment]
Hmm,

I think I banged my head to many times against the wall :(. When I use your example project it works fine. I then checked my project, removed all global installs, did a fresh install and it still didn't work. 

I then checked the plugin code I'm currently writing and I now need to shame myself (sitting in the corner right now with the donkey head on). I had 1 line of code that was cleaning the result folder in the protractor plugin `[setup](https://github.com/angular/protractor/blob/master/lib/plugins.ts#L39)` phase.

According to the documentation the _WebDriver session has been started_ but the _test-framework shouldn't been set up_. It looks like this is wrong because when the WebDriver session has been started Cucumber has already been loaded. This means that Cucumber already did the `fs.open(path.resolve(this.cwd, outputTo), 'w')` meaning that the file is already created. My code removed that folder after the file had been created. The second feature file that created a json file with a PID was pulled up created a file looking like it was clobbering with the previous one.

Sorry for bothering you with this and thanks for helping me out with this
